mdv/mga's changes:
- [2011-07-14] add a EXCLUDE_FROM_FULL_STRIP environment variable so that we
  can exclude files from being completely stripped, but for their debugging symbols
- [2010-03-09] allow for only stripping files without creating debug
  package contents by setting DISABLE_DEBUG environment variable
- [2010-03-23] really check that the debug directory we want to search through
  exists, not only the parent

--- a/scripts/find-debuginfo.in
+++ b/scripts/find-debuginfo.in
@@ -410,8 +410,9 @@
   $strip_glibs && case "$(file -bi "$2")" in
     application/x-sharedlib*) g=-g ;;
   esac
-  eu-strip --remove-comment $r $g ${keep_remove_args} -f "$1" "$2" || exit
-  chmod 444 "$1" || exit
+  [ -n "$EXCLUDE_FULL_REGEXP" ] && grep -E -q "$EXCLUDE_FULL_REGEXP" <<< "$2" && g=-g
+  eu-strip --remove-comment $r $g ${keep_remove_args} $([ -n "$DISABLE_DEBUG" ] || echo -f "$1") "$2" || exit
+  [ -n "$DISABLE_DEBUG" ] || chmod 444 "$1" || exit
 }
 
 add_minidebug()
@@ -535,6 +536,12 @@
   echo "$nlinks $inum $f" >>"$temp/primary"
 done
 
+[[ -n "$EXCLUDE_FROM_STRIP" ]] && \
+EXCLUDE_REGEXP=`perl -e 'print "(", join("|", @ARGV), ")"' $EXCLUDE_FROM_STRIP`
+[[ -n "$EXCLUDE_FROM_FULL_STRIP" ]] && \
+EXCLUDE_FULL_REGEXP=`perl -e 'print "(", join("|", @ARGV), ")"' $EXCLUDE_FROM_FULL_STRIP`
+
+echo $EXCLUDE_REGEXP
 # Check ELF or AR file with eu-elflint if --elf-check is given.
 # Returns 1 if --elf-check-strict is given and eu-elflint reports issues.
 # Otherwise returns 0.
@@ -673,6 +680,10 @@
 {
   local nlinks=$1 inum=$2 f=$3 id link linked
 
+  [[ -n "$EXCLUDE_REGEXP" ]] && grep -E -q "$EXCLUDE_REGEXP" <<< "$f" && \
+  return
+  [ -n "$DISABLE_DEBUG" ] && strip_to_debug "" "$f" && return
+
   # reject files already located under /usr/lib/debug, presumably processed
   get_debugfn "$f"
   [ -f "${debugfn}" ] && return 0
@@ -973,7 +984,7 @@
 
 if [ -d "${RPM_BUILD_ROOT}/usr/lib" ] || [ -d "${RPM_BUILD_ROOT}/usr/src" ]; then
   ((nout > 0)) ||
-  test ! -d "${RPM_BUILD_ROOT}/usr/lib" ||
+  test ! -d "${RPM_BUILD_ROOT}/usr/lib/debug" ||
   (cd "${RPM_BUILD_ROOT}/usr/lib"; find debug -type d) |
   sed 's,^,%dir /usr/lib/,' >> "$LISTFILE"
 
